src/libostree/ostree-repo-pull.c

```diff
@@ -1435,7 +1435,7 @@ enqueue_one_object_request (OtPullData        *pull_data,
   expected_max_size_p = (fetchtype != OSTREE_FETCH_OBJECT_CORE) ? NULL : g_hash_table_lookup (pull_data->expected_commit_sizes, checksum);
   if (expected_max_size_p)
     expected_max_size = *expected_max_size_p;
-  else if (is_meta)
+  else if (is_meta && fetchtype != OSTREE_FETCH_OBJECT_COMPAT_SIZES)
```

Comment:
It gets assigned to `OSTREE_MAX_METADATA_SIZE` if `fetchtype != OSTREE_FETCH_COMPAT_OBJECT_SIZES`. Otherwise, it falls through to `0`, which gets interpreted as unlimited size.